### PR TITLE
Add version APIs for the driver contracts

### PR DESCRIPTION
**Pull request by kvinwang**
Changes:
- The major change in this PR is adding version APIs for the driver contracts.
- The System::version() returns a tuple of 3 elements rather than 2. (This is a breaking change in theory, but should not break in practice. I think it should be rarely used externally)
- Added `start/stop` sidevm for the log_server contract. In case we want to upgrade the log server, we want to stop the old sidevm.
- Added `set_code` API for `sidevm_deployer`. Because `sidevm_deployer` has some state, we add this API to support in-place upgrading. This API is a bit crude, for example, it doesn't check the compatibility, migrate states .etc. so we need to be careful to design the future upgrade flow. The other two drivers(log_server, tokenomic) are stateless, so they don't need a set_code API and we can just deploy a new instance for the newer version and link the driver to the new instance.

We should launch the first cluster with changes in this PR. @h4x3rotab @shelvenzhou

### CODE REVIEW

File: crates/pink-drivers/tokenomic/lib.rs [hunk 1]
```diff
@@ -26,6 +26,11 @@ mod tokenomic {
             }
             Err(Error::BadOrigin)
         }
+
+        #[ink(message)]
+        pub fn version(&self) -> this_crate::VersionTuple {
+            this_crate::version_tuple!()
```
Comment (h4x3rotab) [inline crates/pink-drivers/tokenomic/lib.rs:32]: Very elegant solution.

File: crates/pink-drivers/log_server/lib.rs [hunk 2]
```diff
@@ -10,14 +10,33 @@ mod contract {
...
             Self {}
         }
 
+        #[ink(message)]
+        pub fn version(&self) -> this_crate::VersionTuple {
+            this_crate::version_tuple!()
+        }
+
+        #[ink(message)]
+        pub fn start(&self) {
```
Comment (h4x3rotab) [inline crates/pink-drivers/log_server/lib.rs:31]: Anyone can call?

Comment (kvinwang) [reply crates/pink-drivers/log_server/lib.rs]: Added permission check, and added a `#[ink(message)] fn owner()` for each driver.

Comment (h4x3rotab) [reply crates/pink-drivers/log_server/lib.rs]: Is owner the system contract or the developer who deployed it?

Comment (kvinwang) [reply crates/pink-drivers/log_server/lib.rs]: The one who deployed it. The system contract already has permission to stop any contract's sidevm.

**Comment by kvinwang** [comment]
@h4x3rotab Can we merge this now?

**Review by Leechael — APPROVED** [review]
lgtm

**Comment by kvinwang** [comment]
> One quick question (but maybe is what next), can we force each contract MUST implement the `version` method?

I don't think we should do so, because we support vanilla ink contracts.

**Comment by Leechael** [comment]
> > One quick question (but maybe is what next), can we force each contract MUST implement the `version` method?
> 
> I don't think we should do so, because we support vanilla ink contracts.

Make sense. We can introduce it as a best practice and provide additional support on the UI to encourage this pattern.

**Comment by h4x3rotab** [comment]
> > > One quick question (but maybe is what next), can we force each contract MUST implement the `version` method?
> > 
> > 
> > I don't think we should do so, because we support vanilla ink contracts.
> 
> Make sense. We can introduce it as a best practice and provide additional support on the UI to encourage this pattern.

On mature community this is exactly how the standard works. It's not enforced but widely adopted by the community. The best case is Ethereum's [ERC20](https://eips.ethereum.org/EIPS/eip-20) standard. The same idea exists in the ink! community as well, but in terms of the implementation, it's a little bit harder than it in Solidity (rust doesn't have inherit). Samples can be found in [the official docs](https://use.ink/basics/trait-definitions).

**Review by h4x3rotab — APPROVED** [review]
LGTM

**Pull request merged**